src/Extension/FrontMatter/FrontMatterParser.php

```diff
@@ -0,0 +1,63 @@
+<?php
+
+declare(strict_types=1);
+
+/*
+ * This file is part of the league/commonmark package.
+ *
+ * (c) Colin O'Dell <colinodell@gmail.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace League\CommonMark\Extension\FrontMatter;
+
+use League\CommonMark\Event\DocumentPreParsedEvent;
+use League\CommonMark\Extension\FrontMatter\Yaml\FrontMatterParserInterface;
+use League\CommonMark\Input\MarkdownInput;
+use League\CommonMark\Parser\Cursor;
+
+final class FrontMatterParser
+{
+    /**
+     * @var FrontMatterParserInterface
+     *
+     * @psalm-readonly
+     */
+    private $yamlParser;
+
+    private const REGEX_FRONT_MATTER = '/^---\\n.*\\n---\n/s';
+
+    public function __construct(FrontMatterParserInterface $yamlParser)
+    {
+        $this->yamlParser = $yamlParser;
+    }
+
+    public function __invoke(DocumentPreParsedEvent $event): void
+    {
+        $content = $event->getMarkdown()->getContent();
+
+        $cursor = new Cursor($content);
+
+        // Locate the front matter
+        $frontMatter = $cursor->match(self::REGEX_FRONT_MATTER);
+        if ($frontMatter === null) {
+            return;
+        }
+
+        // Trim the last 4 characters (ending ---s and newline)
+        $frontMatter = \substr($frontMatter, 0, -4);
+
+        // Parse the resulting YAML data
+        $data = $this->yamlParser->parse($frontMatter);
+
+        // Store the parsed data in the Document
+        $event->getDocument()->data['front_matter'] = $data;
+
+        // Advance through any remaining newlines which separated the front matter from the Markdown text
+        $cursor->match('/^\n+/');
```

Comment:
I'd still use `'/^\R+/'` here.

Reply:
Newlines are already normalized to `\n` characters at this point thanks to the `MarkdownInput` class:
 https://github.com/thephpleague/commonmark/blob/0125ec4c442e61954b0c4c964ea50448e4067fde/src/Input/MarkdownInput.php#L83

This normalization occurs just before the event we're listening to is fired, thus guaranteeing that we'll only have Linux newlines for the rest of the parsing: 

https://github.com/thephpleague/commonmark/blob/71e723e23f41a9ceac34941cf87953eb3009ce37/src/Parser/MarkdownParser.php#L109